### ISSUE TITLE
consistent-this should have an option to limit it's scope to const values

### ISSUE THREAD

**Issue by paul-go** [eslint/eslint#11482]
**What rule do you want to change?**
consistent-this

**Does this change cause the rule to produce more or fewer warnings?**
Fewer

**How will the change be implemented? (New option, new default behavior, etc.)?**
New option

**Please provide some example code that this change will affect:**

```js
const nodes: X.Node[] = [];
		
let currentLevel = this;
while ((currentLevel = currentLevel.container) !== null)
	nodes.push(currentLevel);
```
**What does the rule currently do for this code?**
As far as I can tell, the consistent-this rule doesn't really make semantic sense in the case when the value is being mutated. In the example I showed, there is an iteration that is passing through multiple levels of container objects, which starts at the current object (this). You wouldn't want said value to be called `self` or `me`. And so I believe an additional option to limit the scope of this rule to only apply to `const` values would allow this rule to be more useful. Any leakage would be mitigated for anyone also using the `prefer-const` rule.

**Are you willing to submit a pull request to implement this change?**
Possibly

**Comment by platinumazure** [comment]
Hi @paul-go, thanks for the issue and sorry for letting this slip through the cracks.

Personally, I think this is enough of a corner case that it would make the most sense for you to use an ESLint inline disable comment, to reflect that the rule doesn't really make sense there. I'm not sure it's worth adding a new option to the rule for this case. But I would be open to the possibility if another team member believes it is worthwhile.

Using disable comments in your case:

```js
const nodes: X.Node[] = [];
		
let currentLevel = this;   // eslint-disable-line consistent-this
while ((currentLevel = currentLevel.container) !== null)
	nodes.push(currentLevel);
```
